Switch to CircleCI 2.0 and add test of example-wordpress-composer

stevector opened the pull request:
I expect that Circle 2.0 will make it easier to test multiple example repos.

stevector commented:
@greg-1-anderson, I won't be able to make progress on this tomorrow but I do have time Friday. If you have availability tomorrow could you look at the configuration of Fan Out/Fan In https://circleci.com/docs/2.0/workflows/#fan-outfan-in-workflow-example and try it with D8 and WordPress?

greg-1-anderson commented:
I fixed the `circle watch` thing by putting it on the same line as the `circle token` command in 34072a8. Frankly, I'm not sure why it didn't work the other way, but I'm glad that it's working like this.

greg-1-anderson commented:
I made a do-nothing final Circle job that is supposed to run after test_wordpress and test_drupal_8 complete. The test_drupal_8 spuriously failed, and then succeeded after I re-ran it. Looks like the final job did not run. Interesting.

stevector commented:
Thanks for pushing this forward @greg-1-anderson! I just added a commit to pass in `SOURCE_COMPOSER_PROJECT` to `test-example-repo.sh` as an argument.

stevector commented:
And we are back to green here! We should get https://github.com/pantheon-systems/example-drops-8-composer/issues/95 merged next

stevector commented on .circleci/config.yml:
```diff
@@ -0,0 +1,69 @@
+                command: ./.circleci/set-up-globals.sh
+            - run:
+                name: Test D8
+                command: ./.circleci/test-example-repo.sh "pantheon-systems/example-drops-8-composer:dev-git-ignore-file-perms"
```
I'd like to switch back to `pantheon-systems/example-drops-8-composer:dev-master`

greg-1-anderson replied:
Note that we have a `test` and a `test-2.x` branch to ensure that we [install the same version of the build tools plugin in the generated environment as in the original environment](https://github.com/pantheon-systems/example-drops-8-composer/blob/test/circle.yml#L50).

stevector replied:
@greg-1-anderson ok, it looks like test-2.x is a few commits behind master. Is it just manual merging to keep those up to date?

greg-1-anderson replied:
Yeah; I should probably have some process to update those branches automatically.

stevector commented on .circleci/config.yml:
```diff
@@ -0,0 +1,69 @@
+            - run:
+                name: Test WordPress
+                # @todo, change to pantheon-systems
+                command: ./.circleci/test-example-repo.sh "stevector/example-wordpress-composer:dev-master"
```
`stevector` should switch to `pantheon-systems`. See https://github.com/pantheon-systems/example-wordpress-composer/issues/1

stevector commented:
My last two commits set the packages to be tested as `"pantheon-systems/example-drops-8-composer:dev-test-2.x"` and `"pantheon-systems/example-wordpress-composer:dev-master"`. @greg-1-anderson and I just talked about this PR and agreed that we can merge once we see passing tests again. We both expect a lot of revision and improvements on our usage of Circle 2.0 in follow up PRs.

stevector commented:
@greg-1-anderson and somehow we are back to merge error on file permissions for the d8 build.

greg-1-anderson commented:
Huh? That is really odd. Is someone setting the `x` bit on some files? On all files? Where is that happening? Why doesn't your git config setting mask the problem? Why isn't this a problem for the WordPress tests?

stevector commented:
@greg-1-anderson The conflict is not on all files. Just a subset. 

```
~/Sites/build-tools-325   master ●✚ >M<  git status
On branch master
Your branch is up-to-date with 'origin/master'.
You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:

	modified:   .gitignore
	modified:   build-metadata.json
	modified:   vendor/autoload.php
	modified:   vendor/composer/autoload_classmap.php
	modified:   vendor/composer/autoload_real.php
	modified:   vendor/composer/autoload_static.php

Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both added:      vendor/asm89/stack-cors/src/Asm89/Stack/Cors.php
	both added:      vendor/asm89/stack-cors/src/Asm89/Stack/CorsService.php
	both added:      vendor/asm89/stack-cors/test/Asm89/Stack/CorsTest.php
	both added:      vendor/behat/gherkin/bin/update_i18n
	both added:      vendor/behat/mink-extension/README.md
	both added:      vendor/behat/mink-extension/build.php
	both added:      vendor/behat/mink-selenium2-driver/src/Selenium2Driver.php
	both added:      vendor/doctrine/instantiator/.travis.install.sh
	both added:      vendor/drupal/console-en/translations/module.uninstall.yml
	both added:      vendor/drupal/console/src/Command/Module/UninstallCommand.php
	both added:      vendor/drush/drush/examples/drush.wrapper
	both added:      vendor/drush/drush/examples/git-bisect.example.sh
	both added:      vendor/drush/drush/examples/helloworld.script
	both added:      vendor/drush/drush/misc/windrush_build/windrush_build
	both added:      vendor/drush/drush/tests/testChildren.sh
	both added:      vendor/drush/drush/unish.sh
	both added:      vendor/gabordemooij/redbean/README.markdown
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Adapter.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Adapter/DBAdapter.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/BeanHelper.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/BeanHelper/SimpleFacadeBeanHelper.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Driver.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Driver/RPDO.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Facade.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Logger.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Logger/RDefault.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/OODB.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/OODBBean.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Observable.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/Observer.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/QueryWriter.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/QueryWriter/AQueryWriter.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/QueryWriter/MySQL.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/QueryWriter/PostgreSQL.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/QueryWriter/SQLiteT.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/RedException.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/RedException/SQL.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/SimpleModel.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/SimpleModelHelper.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/ToolBox.php
	both added:      vendor/gabordemooij/redbean/RedBeanPHP/license.txt
	both added:      vendor/gabordemooij/redbean/run_all_tests.sh
	both added:      vendor/gabordemooij/redbean/run_single_test.sh
	both added:      vendor/gabordemooij/redbean/test-dist.ini
	both added:      vendor/instaclick/php-webdriver/lib/WebDriver/Service/CurlService.php
	both added:      vendor/instaclick/php-webdriver/lib/WebDriver/Service/CurlServiceInterface.php
	both added:      vendor/instaclick/php-webdriver/lib/WebDriver/ServiceFactory.php
	both added:      vendor/jcalderonzumba/gastonjs/bin/run-tests.sh
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/static/auth_ok.html
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/static/basic.html
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/static/fibonacci.js
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/blank.gif
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/bottom.png
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/form.html
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/shadow.gif
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/top.png
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/view.css
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_execute/view.js
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_extensions/blank.gif
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_extensions/bottom.png
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_extensions/form.html
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_extensions/shadow.gif
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_extensions/top.png
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/script_extensions/view.css
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/blank.gif
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/bottom.png
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/form.html
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/shadow.gif
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/top.png
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/view.css
	both added:      vendor/jcalderonzumba/gastonjs/tests/unit/Server/www/web/test/standard_form/view.js
	both added:      vendor/jcalderonzumba/mink-phantomjs-driver/bin/run-tests.sh
	both added:      vendor/nikic/php-parser/test_old/run-php-src.sh
	both added:      vendor/paragonie/random_compat/build-phar.sh
	both added:      vendor/pear/console_table/Table.php
	both added:      vendor/psy/psysh/bin/build
	both added:      vendor/psy/psysh/bin/build-manual
	both added:      vendor/psy/psysh/bin/build-phar
	both added:      vendor/psy/psysh/bin/build-vendor
	both added:      vendor/psy/psysh/bin/build-vendor-compat
	both added:      vendor/psy/psysh/bin/package
	both added:      vendor/psy/psysh/test/tools/gen_unvis_fixtures.py
	both added:      vendor/psy/psysh/test/tools/vis.py
	both added:      vendor/symfony/dom-crawler/Tests/CrawlerTest.php
	both added:      vendor/symfony/finder/Tests/GlobTest.php
	both added:      vendor/zendframework/zend-escaper/doc/book/index.md
	both added:      vendor/zendframework/zend-feed/doc/book/index.md
	both added:      vendor/zendframework/zend-stdlib/doc/book/index.md
	both added:      web/core/scripts/drupal.sh
	both added:      web/core/scripts/password-hash.sh
	both added:      web/core/scripts/rebuild_token_calculator.sh
	both added:      web/core/scripts/run-tests.sh
	both added:      web/core/scripts/update-countries.sh
```

stevector commented:
I'm trying this: https://github.com/pantheon-systems/terminus-build-tools-plugin/pull/60/commits/4eba7e8b9db4cc70c59b25b4f51f87c3ff30153a

stevector commented:
Well, all that did was result in an non-executable drush launcher

stevector commented:
@greg-1-anderson, Circle 2 is running as root.

```
Do not run Composer as root/super user! See https://getcomposer.org/root for details
```

I wonder if that is the problem

stevector commented:
@greg-1-anderson I was able to get the merge to work locally with `ours` as the strategy rather than `theirs` as the `-X` option of the `recursive` strategy.

Let's see if it works.

stevector commented:
I think we know now that 1.x and 2.x of the build plugin are handling the build artifact branches in different ways (that I don't fully understand) and that is what results in the merge conflict.

stevector commented on src/Commands/EnvMergeCommand.php:
```diff
@@ -65,10 +65,11 @@ public function mergeBuildEnv($site_env_id, $options = ['label' => '', 'delete'
+        // @todo, is a merge even necessary when the master branch will be force pushed over?
         $this->passthru('git fetch pantheon');
-        $this->passthru('git checkout pantheon/master');
+        $this->passthru('git checkout pantheon/' . $env_label);
```
@greg-1-anderson Can I get your thoughts on the changes in this file?

greg-1-anderson replied:
Yeah, that seems equivalent to me -- if it works, ship it.

ataylorme commented on src/Commands/EnvMergeCommand.php:
```diff
@@ -65,10 +65,11 @@ public function mergeBuildEnv($site_env_id, $options = ['label' => '', 'delete'
+        $this->passthru('git checkout pantheon/' . $env_label);
         $this->passthru("git checkout -B $tmpMergeBranch");
-        $this->passthru("git merge -q -m 'Merge build assets from test $env_label.' -X theirs $env_id");
+        $this->passthru("git merge -q -m 'Merge build assets from test $env_label.' --strategy=ours pantheon/master");
```
@stevector are we expecting changes in `pantheon/master` that are not in the pull request branch? I guess I'm not clear why we are overwriting the `master` branch.

`git merge --strategy=ours` says
> This resolves any number of heads, but the resulting tree of the merge is always that of the current branch head, effectively ignoring all changes from all other branches. It is meant to be used to supersede old development history of side branches. Note that this is different from the -Xours option to the recursive merge strategy.

`master` shouldn't have old development history. If we are completely replacing `master` with the pull request branch then it would be easier to do
```
git checkout master
git reset --hard pantheon/$env_label
git push --force -q pantheon $tmpMergeBranch:master
```

greg-1-anderson replied:
Yeah, I think I like it @ataylorme 's way best. Give it a try.

greg-1-anderson commented:
I'm going to go ahead and merge this, since it's working well enough. I think someone should try @ataylorme 's suggestion in a follow-on PR at some point.
